osu.Game.Rulesets.Catch/Difficulty/Skills/Movement.cs

```diff
@@ -57,49 +66,107 @@ protected override double StrainValueOf(DifficultyHitObject current)
-            double sqrtStrain = Math.Sqrt(weightedStrainTime);
+            double movementValue;
 
-            double edgeDashBonus = 0;
+            // Counting direction changes for length scaling in the performance calculation
+            if (Math.Sign(distanceMoved) != Math.Sign(lastDistanceMoved) && Math.Sign(lastDistanceMoved) != 0 && Math.Abs(distanceMoved) > 0.1)
+            {
+                DirectionChangeCount += 1;
+            }
 
-            // Direction change bonus.
-            if (Math.Abs(distanceMoved) > 0.1)
+            // Separating HyperDashes and basic fruits calculations
+            // Basic fruit calculation
+            if (!catchCurrent.LastObject.HyperDash)
             {
-                if (Math.Abs(lastDistanceMoved) > 0.1 && Math.Sign(distanceMoved) != Math.Sign(lastDistanceMoved))
+                // The base value is a ratio between distance moved and strain time
+                movementValue = 0.071 * Math.Abs(distanceMoved) / weightedStrainTime;
+
+                if (Math.Abs(distanceMoved) > 0.1 && Math.Sign(distanceMoved) != Math.Sign(lastDistanceMoved) && Math.Sign(lastDistanceMoved) != 0)
                 {
-                    double bonusFactor = Math.Min(50, Math.Abs(distanceMoved)) / 50;
-                    double antiflowFactor = Math.Max(Math.Min(70, Math.Abs(lastDistanceMoved)) / 70, 0.38);
+                    // We buff shorter movements upon direction change
+                    movementValue *= 1 + (23 / Math.Pow(Math.Abs(distanceMoved), 0.7));
+                }
+            }
+            else
+            {
+                // Hyperdashes calculation
+                // Both strain time and distance moved are scaled down because both factors are not optimally representing the difficulty
+                movementValue = 0.06 * Math.Pow(Math.Abs(distanceMoved), 0.3) / Math.Pow(weightedStrainTime, 0.3);
+
+                // Direction change scaling
+                if (!isSameDirection && lastStrainTime > 0)
+                    movementValue *= 1 + 0.17 * Math.Pow(Math.Abs(lastExactDistanceMoved), 1.2) / Math.Pow(lastStrainTime, 0.75);
+                else movementValue *= 0.7;
+
+                // Handling hyperdash chains
+               if (previousIsDoubleHdash)
+               {
+                    // Scaling hyperdash chains according to movement
+                    movementValue *= isSameDirection ? Math.Abs(distanceMoved / 60) : 0.5;
+               }
+            }
 
-                    distanceAddition += direction_change_bonus / Math.Sqrt(lastStrainTime + 16) * bonusFactor * antiflowFactor * Math.Max(1 - Math.Pow(weightedStrainTime / 1000, 3), 0);
+            // We handle the case of tap-dashes ending with hdashes (stacks that require tapping in the same direction)
+            if (catchCurrent.LastObject.HyperDash && Math.Abs(distanceMoved) > HalfCatcherWidth && isTapDashBoosted)
```

Comment:
`HalfCatcherWidth` is compared against the normalized distance.

Reply:
I've used that to make sure the distance moved is enough so it doesn't buff platewalk, in most cases, if you have a suggestion for something more consistent, I'm taking it.

DT/HT are not completely ignored, it already changes the StrainTime. So the only reason to use clockRate for me would be for cases like edge dashs or something like that.
 It's hard to tell you exactly why DT is not nerfed and HT is buffed compared to stable because as you can see it's a completely new system that doesn't really behave the same way. I guess some HT are buffed due to low AR buff and DT is not nerfed due to my "normal movement" calculation and the high AR buff.

As for what changes are good or not, your suggestion is that I provide a few maps for each line/block of logic to explain my reasoning, right?

Reply:
(Please post a separated comment for reply not about this particular change)

Probably my comment <https://github.com/ppy/osu/pull/14936#issuecomment-940693540> was not clear enough.
To restate the issue, `NormalizedPosition` is multiplied by a scaling factor <https://github.com/ppy/osu/blob/900549a7713d759b37c69639f420d60f22888f1e/osu.Game.Rulesets.Catch/Difficulty/Preprocessing/CatchDifficultyHitObject.cs#L33> but `HalfCatcherWidth` is the length in the original coordinate <https://github.com/ppy/osu/blob/900549a7713d759b37c69639f420d60f22888f1e/osu.Game.Rulesets.Catch/Difficulty/CatchDifficultyCalculator.cs#L72-L75>
So, if you want to compare a normalized distance with the catcher width, `HalfCatcherWidth` should be multiplied by the scaling factor as well (or use the non-normalized distance).

Reply:
I wasn't aware of this, I'm gonna multiply it by the scaling factor and see how it behaves.
Also, sorry for responding on this comment, I'm not really familiar with pull requests...